dot/parachain/util/util.go

```diff
@@ -34,6 +36,123 @@ type Ancestors struct {
+	BenefitMajorFirst UnifiedReputationChangeType = "BenefitMajorFirst"
+	BenefitMajor      UnifiedReputationChangeType = "BenefitMajor"
+)
+
+// UnifiedReputationChange represents a reputation change for a peer.
+type UnifiedReputationChange struct {
+	Type   UnifiedReputationChangeType
+	Reason string
+}
+
+// CostOrBenefit returns the cost or benefit of the reputation change.
+func (u UnifiedReputationChange) CostOrBenefit() int32 {
+	switch u.Type {
+	case CostMinor:
+		return -100_000
+	case CostMajor:
+		return -300_000
+	case CostMinorRepeated:
+		return -200_000
+	case CostMajorRepeated:
+		return -600_000
+	case Malicious:
+		return -1 << 31 // Equivalent to i32::MIN
+	case BenefitMajorFirst:
+		return 300_000
+	case BenefitMajor:
+		return 200_000
+	case BenefitMinorFirst:
+		return 15_000
+	case BenefitMinor:
+		return 10_000
+	default:
+		return 0
+	}
+}
+
+// ReputationAggregator collects and sends reputation changes in batches.
+type ReputationAggregator struct {
+	sendImmediatelyIf func(rep UnifiedReputationChange) bool
+	byPeer            map[peer.ID]int32
+	mu                sync.Mutex
+}
+
+// NewReputationAggregator creates a new ReputationAggregator.
+func NewReputationAggregator(sendImmediatelyIf func(rep UnifiedReputationChange) bool) *ReputationAggregator {
+	return &ReputationAggregator{
+		sendImmediatelyIf: sendImmediatelyIf,
+		byPeer:            make(map[peer.ID]int32),
+	}
+}
+
+// Send sends the accumulated reputation changes in a batch and clears the state.
+func (r *ReputationAggregator) Send(overseerCh chan<- NetworkBridgeTxMessage) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if len(r.byPeer) == 0 {
+		return
+	}
+
+	message := NetworkBridgeTxMessage{
+		ReportPeerMessageBatch: r.byPeer,
+	}
+	overseerCh <- message
+
+	r.byPeer = make(map[peer.ID]int32)
+}
+
+// Modify processes a reputation change, sending it immediately if necessary or accumulating it.
+func (r *ReputationAggregator) Modify(overseerCh chan<- NetworkBridgeTxMessage, peerID peer.ID, rep UnifiedReputationChange) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.sendImmediatelyIf(rep) {
+		r.singleSend(overseerCh, peerID, rep)
+		return
+	}
+
+	r.add(peerID, rep)
+	fmt.Printf("Accumulated reputation change for peer %s: %+v\n", peerID, rep)
+}
+
+// singleSend sends a single reputation change directly.
+func (r *ReputationAggregator) singleSend(overseerCh chan<- NetworkBridgeTxMessage, peerID peer.ID, rep UnifiedReputationChange) {
+	message := NetworkBridgeTxMessage{
+		ReportPeerMessageBatch: map[peer.ID]int32{
+			peerID: rep.CostOrBenefit(),
+		},
+	}
+	overseerCh <- message
+	fmt.Printf("Sent immediate reputation change for peer %s: %+v\n", peerID, rep)
+}
+
+// add accumulates a reputation change for a peer.
+func (r *ReputationAggregator) add(peerID peer.ID, rep UnifiedReputationChange) {
+	if _, exists := r.byPeer[peerID]; !exists {
+		r.byPeer[peerID] = 0
+	}
+	r.byPeer[peerID] += rep.CostOrBenefit()
```

Comment:
Actually, here we don't need the add function since we can place this code in the `Modify` method. If we get `r.byPeer[peerID]` and the entry for `peerID` does not exists it will return 0 by default, so we don't need to check if it exists or not.

```suggestion
	r.byPeer[peerID] = primitives.SaturatingAdd(r.byPeer[peerID], rep.CostOrBenefit())
```

Reply:
@EclesioMeloJunior fixed